### PR TITLE
feat(metrics): support monitor network connection of S3 and grpc

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In tonic gRPC and S3 client, they both support accepting a connector that is responsible for creating new connections. The connector implements the `tower::Service<Uri>` trait, which takes an endpoint uri and return a new connection to the uri. The connection is not of any concrete type, but any type that implements `AsyncRead` and `AsyncWrite` trait. In tonic service, it support accepting a stream that yields new connection with client rather than a concrete tcp listener. The connections it yields are also not of any concrete type, but still types implement `AsyncRead` and `AsyncWrite`. Therefore, it is feasible for us to implement a monitor wrapper that wraps the default implementation and implements these traits, and while the methods of these traits are called, we can monitor the inner implementation.

In this PR, we implements such wrapper as `MonitoredConnection`. It monitors the read and write throughput, existing count and err rate of the low level tcp connection. The network connection of tonic client and server, and S3 client are monitored in this PR and display on the grafana dashboard.

The newly added panels look like
image

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/monitor/connection.rs [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+
+    monitor_connector(http, connection_type)
+}
+
+pub fn monitor_connector<C>(
+    connector: C,
+    connection_type: impl Into<String>,
+) -> MonitoredConnection<C, MonitorNewConnectionImpl> {
+    let connection_type = connection_type.into();
+    info!(
```
Comment (BugenZhao) [inline src/common/src/monitor/connection.rs:393]: Is this logging intended?

File: src/common/src/monitor/connection.rs [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+where
+    Con: Connected<ConnectInfo = TcpConnectInfo>,
+{
+    type Item = Result<MonitoredConnection<Con, M::ConnectionMonitor>, E>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let (inner, monitor) = MonitoredConnection::project_into(self);
+        inner.poll_next(cx).map(|opt| {
+            opt.map(|result| {
+                result.map(|conn| {
```
Comment (BugenZhao) [inline src/common/src/monitor/connection.rs:257]: Should we also monitor accept error here?

File: src/compute/src/server.rs [hunk 3]
```diff
@@ -417,23 +416,45 @@ pub async fn compute_node_serve(
...
+                    shutdown_signal,
+                )
+                .await
+                .unwrap();
+        }
+        #[cfg(madsim)]
+        server
+            .serve_with_shutdown(listen_addr, shutdown_signal)
             .await
             .unwrap();
```
Comment (BugenZhao) [inline src/compute/src/server.rs:459]: What about extracting this into an extension trait on `Router`, with `easy_ext::ext`?

**Review by BugenZhao — APPROVED** [review]
LGTM

File: src/common/src/monitor/connection.rs (hunk 1 above)
Comment (wenym1) [reply src/common/src/monitor/connection.rs:384]: Yes. The method is only called during initialization, so there won't be too many this log.

File: src/common/src/monitor/connection.rs (hunk 2 above)
Comment (wenym1) [reply src/common/src/monitor/connection.rs:260]: It's like a try stream. On accept error the stream will be stopped, and in upper layer, they server will fail to listen to a port and panic. So I think we don't need to monitor the accept error here.

File: src/compute/src/server.rs (hunk 3 above)
Comment (wenym1) [reply src/compute/src/server.rs]: Great idea! I have changed it.

**Review by BugenZhao — APPROVED** [review]
LGTM!

**Review by fuyufjh — APPROVED** [review]

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**